### ISSUE TITLE
OperationalError: (_mysql_exceptions.OperationalError) (2013, 'Lost connection to MySQL server during query')

### ISSUE THREAD

**Issue by sagor999** [vitessio/vitess#6208]
## Overview of the Issue

We are seeing random connections lost to vitess sql during queries.
All health checks return healthy.

This is an error:
```
OperationalError: (_mysql_exceptions.OperationalError) (2013, 'Lost connection to MySQL server during query')
[SQL: SELECT `Tasks`.tags AS `Tasks_tags`, `Tasks`.output AS `Tasks_output`, `Tasks`.id AS `Tasks_id`, `Tasks`.identifier AS `Tasks_identifier`, `Tasks`.name AS `Tasks_name`, `Tasks`.created AS `Tasks_created`, `Tasks`.ready AS `Tasks_ready`, `Tasks`.started AS `Tasks_started`, `Tasks`.ended AS `Tasks_ended`, `Tasks`.status AS `Tasks_status`, `Tasks`.inherited_status AS `Tasks_inherited_status`, `Tasks`.first_created AS `Tasks_first_created`, `Tasks`.first_ready AS `Tasks_first_ready`, `Tasks`.inherited_id AS `Tasks_inherited_id`, `Tasks`.dependency_identifiers AS `Tasks_dependency_identifiers`, `Tasks`.resource_requirements AS `Tasks_resource_requirements`, `Tasks`.dependencies AS `Tasks_dependencies`, `Tasks`.build_id AS `Tasks_build_id`, `Tasks`.build_id_archived AS `Tasks_build_id_archived`, `Tasks`.buildconfig_id AS `Tasks_buildconfig_id`, `Tasks`.version AS `Tasks_version`, `Tasks`.argshash AS `Tasks_argshash`, `Tasks`.artefactdir AS `Tasks_artefactdir`, `Tasks`.flags AS `Tasks_flags`, `Tasks`.stage AS `Tasks_stage`, `Tasks`.priority AS `Tasks_priority`, `Tasks`.error_id AS `Tasks_error_id`, `Tasks`.scheduler_error AS `Tasks_scheduler_error` 
FROM `Tasks` 
WHERE `Tasks`.id = %s]
[parameters: (53131,)]
```

Deployed into kubernetes (1.17.3) via latest helm chart with default settings.
This is how db is setup:
```
          keyspaces:
            - name: "test"
              shards:
                - name: "0"
                  tablets:
                    - type: "replica"
                      vttablet:
                        replicas: 2
```

Only thing I see in logs is this constant spam:
```
[Note] Aborted connection 9029 to db: 'unconnected' user: 'vt_dba' host: 'localhost' (Got an error reading communication packets)
```
And also this:
```
syntax error at position 33 near ':vtg2' (CallerID: userData1)
: Sql: "select convert(:vtg1, CHAR(:vtg2)) as anon_1 from dual", BindVars: {vtg1: "type:VARBINARY value:\"test plain returns\" "vtg2: "type:INT64 value:\"60\" "}
```

Not sure if it might be relevant or not, but that error keeps spamming, doesn't seem to happen only during disconnect. 

Any help would be greatly appreciated.

**Comment by sougou** [comment]
A 2013 error means that the connection was closed after the query was sent to mysql, which can happen if the query killer intervenes due to a timeout. Another possibility is that mysql closed a connection due to `wait_timeout`, but I believe you'd get a 2006 error in that case.

Can you check `/debug/vars` for the vttablets. See if there are counts against Errors or Kills? Also check the logs of the vttablet in the container.

The other errors you reported seem unrelated. The syntax error needs to be independently investigated.

**Comment by sagor999** [comment]
@sougou 
for errors I see this:
`INVALID_ARGUMENT": 2284`
Nothing for kills. 

for vttablet log: only thing that is in that log is just those three lines, that repeat over and over again:
```
May 20, 2020 @ 13:39:15.578 | syntax error at position 33 near ':vtg2' (CallerID: userData1)

May 20, 2020 @ 13:39:15.578 | : Sql: "select convert(:vtg1, CHAR(:vtg2)) as anon_1 from dual", BindVars: {vtg1: "type:VARBINARY value:\"test plain returns\" "vtg2: "type:INT64 value:\"60\" "}

May 20, 2020 @ 13:39:15.578 | I0520 20:39:15.578246       1 tabletserver.go:1522] Code: INVALID_ARGUMENT
```

**Comment by sougou** [comment]
The error looks like a bug in vitess. We'll have @systay and @harshit-gangal take a look. Can you identify the original source query? It's probably something that contains `convert(xxx, CHAR(yyy))` as a sub-expression.

Another thing you may want to look at is you're exceeding the max connection limits of mysql.

Also piping the output of `/debug/querylog` from the vttablet may give a clue. This should be easy if you have the latest helm that has vtctld reverse proxying the vttablets.

**Comment by sagor999** [comment]
@sougou I will try to find source query for you, thank you!

how do I increase max connection limits of mysql? is that exposed as config or is it hard coded limit?

I used `/debug/querylog` but only converts I see are those:
```
select convert(:vtg1, CHAR(:vtg2)) as anon_1 from dual

select convert(:vtg1, CHAR character set utf8mb4) collate utf8mb4_bin as anon_1 from dual

select convert('test collated returns', CHAR character set utf8mb4) collate utf8mb4_bin as anon_1 from dual limit 10001
```

I will keep you updated once I find actual source query for this.

**Comment by sougou** [comment]
For the querylog, I was looking for more info on the 2013 errors.
For your deploy, make sure you've removed this line from your charts: `deploymentType: test`.
Also, the max mysql connections are likely coming from this file https://github.com/vitessio/vitess/blob/master/config/mycnf/default.cnf#L36, which is probably in the image.

500 connections should be enough. You may also have to tune the vttablet conn pool parameters.

**Comment by sagor999** [comment]
double checked, deployment type is prod:
`deploymentType: prod` 

As for errors, it seems to happen on 3 different sql query, and none of them have `convert`.
Here is 3 of them:
```
(_mysql_exceptions.OperationalError) (2013, 'Lost connection to MySQL server during query')
[SQL: DELETE FROM Builds WHERE id = %s;]
[parameters: (17477,)]
```

```
(_mysql_exceptions.OperationalError) (2013, 'Lost connection to MySQL server during query')
[SQL: SELECT `Tasks`.tags AS `Tasks_tags`, `Tasks`.output AS `Tasks_output`, `Tasks`.id AS `Tasks_id`, `Tasks`.identifier AS `Tasks_identifier`, `Tasks`.name AS `Tasks_name`, `Tasks`.created AS `Tasks_created`, `Tasks`.ready AS `Tasks_ready`, `Tasks`.started AS `Tasks_started`, `Tasks`.ended AS `Tasks_ended`, `Tasks`.status AS `Tasks_status`, `Tasks`.inherited_status AS `Tasks_inherited_status`, `Tasks`.first_created AS `Tasks_first_created`, `Tasks`.first_ready AS `Tasks_first_ready`, `Tasks`.inherited_id AS `Tasks_inherited_id`, `Tasks`.dependency_identifiers AS `Tasks_dependency_identifiers`, `Tasks`.resource_requirements AS `Tasks_resource_requirements`, `Tasks`.dependencies AS `Tasks_dependencies`, `Tasks`.build_id AS `Tasks_build_id`, `Tasks`.build_id_archived AS `Tasks_build_id_archived`, `Tasks`.buildconfig_id AS `Tasks_buildconfig_id`, `Tasks`.version AS `Tasks_version`, `Tasks`.argshash AS `Tasks_argshash`, `Tasks`.artefactdir AS `Tasks_artefactdir`, `Tasks`.flags AS `Tasks_flags`, `Tasks`.stage AS `Tasks_stage`, `Tasks`.priority AS `Tasks_priority`, `Tasks`.error_id AS `Tasks_error_id`, `Tasks`.scheduler_error AS `Tasks_scheduler_error` 
FROM `Tasks` 
WHERE `Tasks`.id = %s]
[parameters: (53131,)]
```

```
OperationalError: (_mysql_exceptions.OperationalError) (2013, 'Lost connection to MySQL server during query')
[SQL: SELECT `Buildconfigs`.id AS `Buildconfigs_id`, `Buildconfigs`.name AS `Buildconfigs_name`, `Buildconfigs`.description AS `Buildconfigs_description`, `Buildconfigs`.project AS `Buildconfigs_project`, `Buildconfigs`.ci AS `Buildconfigs_ci`, `Buildconfigs`.manual AS `Buildconfigs_manual`, `Buildconfigs`.base_id AS `Buildconfigs_base_id`, `Buildconfigs`.up_to_date_time AS `Buildconfigs_up_to_date_time`, `Buildconfigs`.succeeding_time AS `Buildconfigs_succeeding_time`, `Buildconfigs`.failing_time AS `Buildconfigs_failing_time` 
FROM `Buildconfigs` 
WHERE `Buildconfigs`.id = %s]
[parameters: ('SharedMain',)]
```

All 3 queries are pretty simple it seems like.
If it helps, application that talks to sql is written in python. Not sure if that might offer some help or not.

There is also this totally weird error:
```
InterfaceError: (_mysql_exceptions.InterfaceError) (-1, 'error totally whack')
[SQL: SELECT build_id, id, identifier, name, buildconfig_id, status, inherited_id, error_id, inherited_status FROM Tasks WHERE Tasks.build_id IN (18796,18782,18781,18738,18720,18716,18705,18703,18700,18696,18682,18681,18679,18654,18617,18558,18371,18357,18303,18232,18218,17945,17937,17744,17642,17639,17572,17570,17569,17537,17510,17507,17506,17505,17496,17446,17445,17444,17439,17435,17399,17343,17305,17254,17249,17230,17028,17016,16849,16825,16789,16785,16783,16777,16753,16536,16534,16519,16510,16489,16483,16383,16377,16365,16360,16319,16313,16312,16307,16296,16287,16277,16274,16264,16258,16255,16241,16239,16228,16214,16197,16189,16182,16153,16137,16044,15978,15938,15936,15934,15619,15602,15601,15419,15337,15331,15322,15294,15288,15280,15249,15242,15221,15219,15178,15165,15152,15122,15107,15085,15084,15056,15049,15048,15029,15005,14994,14988,14979,14977,14946,14928,14906,14904,14894,14881,14879,14878,14814,14758,14570,14547,14508,14018,12688,12594,11946,11849,11838,11817,11791,11684,11601,11567,11540,11525,11453,11452,11440,11419,11400,11364,11310,11296,11263,11253,11239,11225,11204,11182,11174,11173,11162,11156,11142,11141,11140,11127,11119,11072,11070,11045,11026,11003,11002,10867,10697,10622,10619,10605,10582,10567,10563,10418,10407,10294,10269,10203,10178,10170,10161,10158,10152,10138,10116,10090,10061,10039,10038,10028,10027,9995,9993,9992,9991,9989,9924,9906,9905,9858,9852,9813,9614,9595,9487,9463,9449,9393,9288,9277,9233,9229,9224,9220,9218,9214,9211,9192,9188,9173,9146,9144,9142,9120,9080,9076,8735,8527,8345,8270,8230,8204,8169,8135,8129,8127,8116,8114,8106,8100,8090,8079,8053,8023,8020,8012,8011,7992,7987,7958,7923,7786,7767,7583,7481,7411,7378,7350,7275,7265,7215,7185,7167,7128,7126,7082,7079,7078,7077,7076,7075,7074,7073,7072,7071,7070,7069,7068,7067,7066,7065,7064,7063,7062,7058,6993,6948,6901,6573,6567,6560,6310,5075,3926,3822,3775,3697,3638,3508,3501,3111,2970,2783,2716,2714,2680,2653,2605,2582,2521,2511,2379,2371,2370,2355,2329,2222,2208,2201,2156,1978,1960,1876,1849,1804,1798,1790,1721,1664,1560,1444,1327,1303,1288,1273,1107,1076,1066,1065,1060,928,921,917,874,866,857,829,640,637,628,598,578,490,454,420,391,389,255,239,56,37,7,4)]
```

All 4 errors happen only when running vitess. 
Here is some more config:
```
          mysqlProtocol:
            enabled: true
            authType: none
```
```
vttablet:
  flavor: mysql56
  mysqlSize: prod

vtgate:
  extraFlags:
    mysql_server_version: "8.0.13-Vitess"
```

**Comment by sougou** [comment]
The "whack" error seems to be coming from python client. I see some references on stackoverflow: https://stackoverflow.com/questions/42696621/mysql-json-python-django-error-totally-whack.

The other ones are strange. Were you able to catch them with `/debug/querylog`?

**Comment by sagor999** [comment]
@sougou in which format would it get reported in `/debug/querylog` ?
I tried looking for `2013` but did not find anything useful. 
And it is hard to look through that log as it is just a stream of data.

**Comment by sougou** [comment]
2013 should have been reported as an error in that stream if vttablet saw that failure. I'm wondering if it's a problem with the python client.

Can you see any errors in the vtgate logs? We could also check the `/debug/querylog` of vtgate to see if it reports anything suspicious.

**Comment by sagor999** [comment]
Yeah, there are no mentions of 2013 in both logs (vtgate and vttablet). 
I am going to look at python client. But it is odd that it only started after switching to vitess.

**Comment by sougou** [comment]
can you also check the vtgate logs? maybe the query generated an internal panic which will manifest as a lost connection.

I've also seen this happen if there's a network policy that closes a connection on inactivity, if a query takes longer than expected.

**Comment by harshit-gangal** [comment]
Are you using `sqlalchemy` ? If yes, it does some pre-checks which fails on select cast. Internally Vitess rewrites cast to convert. Test for that [https://github.com/vitessio/vitess/blob/219898c662e59cca8204fd41a6743435f035d481/go/vt/sqlparser/parse_test.go#L1920,L1923]
Fixing that behaviour should work.

**Comment by sagor999** [comment]
Yes, we are using sqlalchemy.

**Comment by harshit-gangal** [comment]
Fixed via #6304
